src/vm/runtimehandles.cpp

```diff
@@ -1360,31 +1360,28 @@ FCIMPL2(FC_BOOL_RET, RuntimeTypeHandle::CanCastTo, ReflectClassBaseObject *pType
     TypeHandle fromHandle = refType->GetType();
     TypeHandle toHandle = refTarget->GetType();
 
-    BOOL iRetVal = 0;
-
-    TypeHandle::CastResult r = fromHandle.CanCastToNoGC(toHandle);
-    if (r == TypeHandle::MaybeCast)
-    {
-        HELPER_METHOD_FRAME_BEGIN_RET_2(refType, refTarget);
-        iRetVal = fromHandle.CanCastTo(toHandle);
-        HELPER_METHOD_FRAME_END();
-    }
-    else
+    TypeHandle::CastResult r = fromHandle.CanCastToCached(toHandle);
+    if (r != TypeHandle::MaybeCast)
     {
-        iRetVal = (r == TypeHandle::CanCast);
+        FC_RETURN_BOOL((BOOL)r);
     }
 
-    // We allow T to be cast to Nullable<T>
-    if (!iRetVal && Nullable::IsNullableType(toHandle) && !fromHandle.IsTypeDesc())
+    BOOL iRetVal;
+    HELPER_METHOD_FRAME_BEGIN_RET_2(refType, refTarget);
     {
-        HELPER_METHOD_FRAME_BEGIN_RET_2(refType, refTarget);
-        if (Nullable::IsNullableForType(toHandle, fromHandle.AsMethodTable())) 
+        // We allow T to be cast to Nullable<T>
+        if (!fromHandle.IsTypeDesc() && Nullable::IsNullableForType(toHandle, fromHandle.AsMethodTable()))
         {
+            // do not put this in the cache (see TypeHandle::CanCastTo and ObjIsInstanceOfCore). 
             iRetVal = TRUE;
         }
-        HELPER_METHOD_FRAME_END();
+        else
+        {
+            iRetVal = fromHandle.CanCastTo(toHandle);
```

Comment:
This is going to do the cache lookup immediately again. Would it make sense to inline the next level here to avoid it?

Reply:
I had implemented that but then reverted for simplicity. We assume the slow path is rare anyways. (whole premise of winning through caching requires that we mostly read from cache and not go here). 
It is not a big deal, but if can be avoided, why not.

I just was unsure about diminishing returns since this is for reflection, i think. So there could be bigger overheads.

It is easy to avoid extra lookup here.

Reply:
>  this is for reflection

A common reflection usage is one-time only because the result is "cached" by whoever is using the reflection. 

I was stepping through this and was surprised how much redundant code runs on the fallback path. I think it would be worth it to streamline it a bit.

Reply:
Inlined the relevant parts of the next level.